Implement OpenStack provisioner with volume types.

jsafrane opened the pull request:
OpenStack Cinder volume types represent generic "type of storage". Cinder
volumes from different Cinder backends (e.g. NetApp or EMC) will get different
volume type.

This volume  type can represent also QoS (e.g. "slow" vs. "fast" volumes from
NetApp).

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21979)

<!-- Reviewable:end -->

jsafrane commented:
cc: @kubernetes/rh-storage

rootfs commented on pkg/volume/cinder/cinder.go:
```diff
@@ -143,13 +145,38 @@ func (plugin *cinderPlugin) NewProvisioner(options volume.VolumeOptions) (volume
+
+		err := json.Unmarshal(bytes, &opts)
+		if err != nil {
+			return opts, fmt.Errorf("error parsing configuration of %s provisioning plugin: %v", plugin.Name(), err)
```
also print `optStr` in the error message

jsafrane replied:
I don't know, it can be pretty long JSON string. I am not fan of code in error messages.

rootfs commented on pkg/volume/cinder/cinder.go:
```diff
@@ -143,13 +145,38 @@ func (plugin *cinderPlugin) NewProvisioner(options volume.VolumeOptions) (volume
+		opts.VolumeType = strings.ToLower(opts.VolumeType)
+	}
+
+	return opts, nil
```
so `opts` could be "" if `optStr` is "", would that be a problem when calling cinder CreateVolume?

jsafrane replied:
No, it will create a default volume.

rootfs commented:
LGTM, thanks.

rootfs commented:
@eparis ^^^

jsafrane committed:
Implement OpenStack provisioner with volume types.

OpenStack Cinder volume types represent generic "type of storage". Cinder
volumes from different Cinder backends (e.g. NetApp or EMC) will get different
volume type.

This volume  type can represent also QoS (e.g. "slow" vs. "fast" volumes from
NetApp).

childsb commented:
This will need a rebase and a refactor to support the upcoming DP changes.  Targeting for 1.4

jsafrane commented:
I (or @rootfs) will post a new one, interface between the controller and volume plugin has changed.
